### ISSUE TITLE
package api  import ( 	"fmt" 	"testing"  	"encoding/json"  	"github.com/stretchr/testify/assert" )  func TestOpString(t *testing.T) {  	s := OptString{}  	b, err := json.Marshal(s)  	fmt.Printf("%s\n", b) 	fmt.Printf("%+v\n", err)  	assert.NoError(t, err)  }[SUBJECT]: [DESCRIPTION]

### ISSUE THREAD

**Issue by exu** [ogen-go/ogen#660]
### What version of ogen are you using?

v0.54.1

### Can this issue be reproduced with the latest version?

Yes just try to run this test in directory where API is generated with OptString

```go
package api

import (
	"testing"

	"encoding/json"
)

func TestOpString(t *testing.T) {

	s := OptString{}

	_, err := json.Marshal(s)

	if err != nil {
		t.Fatalf("error should be nil, got %v", err)
	}

}
```

So for most use cases with fields set to not required when trying to marshal it (e.g. send through event bus) it's not possible to use ogen.

**Comment by ernado** [comment]
Encoding and decoding with `encoding/json` is not supported for `OptString` because following limitations: https://github.com/golang/go/issues/11939

So it is impossible to represent `OptString` with `encoding/json` (pointers are used instead), that's why we are using `go-faster/jx` and custom code-generated json marshaling and unmarshaling.

**Comment by exu** [comment]
@ernado thanks for quick response, so how should I deal with it, can I omit OptString and use pointers like most of other generators do?

**Comment by exu** [comment]
Looks like just handling zeroed value works 

Would it be a case to just do PR to the generator for `Opt*` structs? 

```go 
// MarshalJSON implements stdjson.Marshaler.
func (s OptString) MarshalJSON() ([]byte, error) {
	if !s.Set {
		return []byte(`{}`), nil
	}
	e := jx.Encoder{}
	s.Encode(&e)
	return e.Bytes(), nil
}


```

**Comment by ernado** [comment]
What are you trying to do?

Using `MarshalJSON` on requests, responses or most generated types should work.

Returning `{}` is invalid, probably we should not generate `MarshalJSON` for `OptString` at all, it is impossible to express omitempty struct with `encoding/json`.

**Comment by exu** [comment]
I try to send generated struct through external library (NATS) and don't have control over marshaller

**Comment by exu** [comment]
It's more about generating valid JSON - as looks like currently, we're ending with something like `field: ` when the struct is uninitialized. It would be ok for me to just store pass it (without handling `omitempty`) I just need valid JSON and currently it's not.

**Comment by tdakkota** [comment]
I see there several ways to fix it:

1. Do not generate `MarshalJSON`/`UnmarshalJSON` for the `Opt`/`NilOpt` types.
    * Likely breaks existing code.
    * These types would be marshaled/unmarshaled by `encoding/json`, as JSON objects.

2. Return `[]byte("null")`.
    * Does not follow the [optionals semantic](https://ogen.dev/docs/types/optional#semantic).

3. Return an error in case if value is not set, add `omitempty` tag to optional fields.
    * Forces to add `omitempty` tag. Some people may use generated types
      in their own structs.
    * Would not work properly in some cases. For example: `Set` is `false`, but value is not zero.

**Comment by ernado** [comment]
@tdakkota `omitempty` does not work for structures, even if they are zero value.

**Comment by ernado** [comment]
The only valid solution is (1).

**Comment by exu** [comment]
Maybe just make no 1 optional?

**Comment by art-frela** [comment]
Hi,

for objects workaround with overhead

```go
...
// use generated method MarshalJSON.
tempBts, err := s.MarshalJSON()
if err != nil {
    return fmt.Errorf("change marshal: %w", err)
}

var m map[string]interface{}

// unmarshal to map[string]interface{}
if er := json.Unmarshal(tempBts, &m); er != nil {
    return fmt.Errorf("umarshal to map: %w", er)
}

// use  std MarshalIndent for get ident json.
bts, err := json.MarshalIndent(m, "", "    ")
if err != nil {
    return fmt.Errorf("marshal: %w", err)
}
```



P.S. If all of the Opt fields filled, error not happen!
P.P.S. `ogen --version // ogen version v0.72.1 (built with go1.21.0) darwin/amd64`

**Comment by ernado** [comment]
Yes, we avoid pointers for better semantics and removing NPE problem.

Will use `omitzero` when released tho. Finally.

**Comment by koblas** [comment]
We've got Go 1.24 and `omitzero` can we look at having this patched.

**Comment by kristofferremback** [comment]
Any chance this is being looked at? Otherwise I'd be happy to try help out on #1427 if that's the way to go?

We've had production bug recently due to incorrect usage of `NewOpt[Type](...)`:
```diff
// nils.Unwrap(...) returns the zero value of the passed in pointer type for nil values
- out.Details = ogen.NewOptOrderDetails(nils.Unwrap(orderDetails))
+ if orderDetails != nil {
+ 	out.Details = ogen.NewOptOrderDetails(*orderDetails)
+ }
```

The old code would end up returning something like this (prettified) invalid json:
```json
{
  "id": "9ad94a36-e350-464a-b46c-3e2577dd1193",
  "status": "APPROVED",
  "customerId": "26b6c9e0-454d-4146-a8bd-469121ec0382",
  "details":
  "createdAt": "2025-05-06T07:44:00Z",
  "updatedAt": "2025-05-06T07:44:00Z",
  "fulfilledAt": "2025-05-06T07:44:00Z"
}
```

While the error definitely lies on our side and was an easy fix, it feels like a pretty easy footgun.

**Comment by nonchan7720** [comment]
The problem occurs not only in Marshal but also in Unmarshal.
